include/graphblas/base/blas3.hpp

```diff
@@ -442,6 +442,239 @@ namespace grb {
+	 * @tparam MaskType  The type of the elements in the supplied ALP/GraphBLAS
+	 *                   matrix \a mask.
+	 *
+	 * @param[in, out] x   The result of the reduction.
+	 * 					   Prior value will be considered.
+	 * @param[in]      A   Any ALP/GraphBLAS matrix, will be reduced into \a x.
+	 * @param[in]   mask   Any ALP/GraphBLAS matrix, will mask the matrix \a A.
+	 * 					   Dimensions must match those of \a A.
+	 * @param[in] monoid   The monoid under which to perform this reduction.
+	 * 					   An identity element must be provided when using
+	 * 					   threads in order to perform the local reductions.
+	 *
+	 * @return grb::SUCCESS  When the call completed successfully.
+	 * @return grb::MISMATCH If a \a mask was not empty and does not have size
+	 *                       equal to \a y.
+	 *
+	 * @see grb::foldl provides similar in-place functionality, but folds in a
+	 * 	left-to-right direction.
+	 * @see The same primitive but unmasked is also provided.
+	 *
+	 * \parblock
+	 * \par Valid descriptors
+	 * - descriptors::no_operation: the default descriptor.
+	 * - descriptors::no_casting: the first domain of
+	 * 	 	\a monoid must match \a InputType, the second domain of \a op
+	 * 		match \a IOType, the third domain must match \a IOType.
+	 * - descriptors::transpose_left: A^T will be considered instead 
+	 * 	 	of \a A.
+	 * - descriptors::transpose_right: mask^T will be considered 
+	 * 	 	instead of \a mask.
+	 * - descriptors::invert_mask: Not supported yet.
+	 *
+	 * \note Invalid descriptors will be ignored.
+	 *
+	 * \endparblock
+	 *
+	 * \par Performance semantics
+	 * Each backend must define performance semantics for this primitive.
+	 *
+	 * @see perfSemantics
+	 */
+	template<
+		Descriptor descr = descriptors::no_operation,
+		class Monoid,
+		typename InputType, typename IOType, typename MaskType,
+		typename RIT_A, typename CIT_A, typename NIT_A,
+		typename RIT_M, typename CIT_M, typename NIT_M,
+		Backend backend
+	>
+	RC foldr(
+		IOType &x,
+		const Matrix< InputType, backend, RIT_A, CIT_A, NIT_A > &A,
+		const Matrix< MaskType, backend, RIT_M, CIT_M, NIT_M > &mask,
+		const Monoid &monoid = Monoid(),
+		const typename std::enable_if< !grb::is_object< IOType >::value &&
+			!grb::is_object< InputType >::value &&
+			!grb::is_object< MaskType >::value &&
+			grb::is_monoid< Monoid >::value, void
+		>::type * const = nullptr
+	) {
+#ifndef NDEBUG
+		const bool should_not_call_base_scalar_masked_matrix_foldr = false;
+		assert( should_not_call_base_scalar_masked_matrix_foldr );
+#endif
+		(void) A;
+		(void) x;
+		(void) mask;
+		(void) monoid;
+		return UNSUPPORTED;
+	}
+
+	/**
+	 * Reduces, or \em folds, a matrix into a scalar. 
+	 * Right-to-left unmasked variant.
+	 * 
+	 * Please see the masked grb::foldr variant for a full description.
+	 * 
+	 * \parblock
+	 * 
+	 * \par Valid descriptors specific to this variant
+	 * - descriptors::transpose_matrix: A^T will be considered instead 
```

Comment:
Valid descriptors specific **to this variant**

Reply:
I guess we interpret that differently;)

The way I interpret it, is that what follows, is a list of descriptors that can be given to the current variant. This list is only given for variants that allows descriptors that are not in the base variant OR in the case that it does not allow for descriptors that are in the base variant.

The rationale for providing a full list is that it would be too cumbersome for readers of a reference document to only define what is added to or missing from the descriptors of the base variant.

Reply:
I agree, do you have any suggestion regarding the documentation before I copy it to all the variants ?

Reply:
OK. So I would suggest listing all descriptors that are valid for this specific variant (and similarly elsewhere), including descriptors::no_operation

Reply:
Sure, I meant suggestions for the rest of the documentation